src/main/java/org/broadinstitute/hellbender/tools/spark/BaseRecalibratorSparkOptimized.java

```diff
@@ -0,0 +1,205 @@
+                : IntervalUtils.getAllIntervalsForReference(readsHeader.getSequenceDictionary());
+
+        Stopwatch hacking = Stopwatch.createUnstarted();
+        List<String> localVariants = knownVariants;
+        if (BucketUtils.isCloudStorageUrl(knownVariants.get(0))) {
+            // TODO: fix
+            logger.info("(HACK): copying the GCS file to local just so we can read it back.");
+            hacking.start();
+            localVariants = new ArrayList<>();
+            // this only works with the API_KEY, but then again it's a hack so there's no point in polishing it.
+            PipelineOptions popts = auth.asPipelineOptionsOldStyle();
+            int i=0;
+            for (String s : knownVariants) {
+                // presumably they're all GCS.
+                String d = IOUtils.createTempFile("knownVariants-"+i,".vcf").getAbsolutePath();
+                try {
+                    BucketUtils.copyFile(s, popts, d);
+                } catch (IOException x) {
+                    throw new UserException.CouldNotReadInputFile(s,x);
+                }
+                i++;
+                localVariants.add(d);
+            }
+            hacking.stop();
+            logger.info("Copying the vcf took "+hacking.elapsed(TimeUnit.MILLISECONDS)+" ms.");
+        }
+
+        List<Variant> variants = VariantsSource.getVariantsList(localVariants);
+
+        List<SimpleInterval> shardedIntervals = IntervalUtils.cutToShards(intervals, bigShardSize);
+        ArrayList<ContextShard> localShards = AddContextDataToReadOptimizedFn.fillVariants(shardedIntervals, variants, margin);
+
+        logger.info("Shipping "+localShards.size()+" big shards.");
+        Stopwatch shipping = Stopwatch.createStarted();
+        JavaRDD<ContextShard> shards = ctx.parallelize(localShards);
+        System.out.println("Shipping took: "+shipping.elapsed(TimeUnit.MILLISECONDS) + " ms.");
+
+        logger.info("Loading reads");
+        JavaRDD<ContextShard> reads;
+        try {
+            reads = shards.flatMap( AddContextDataToReadOptimizedFn.subdivideAndFillReads(bam, auth, smallShardSize, margin, readFilterToApply) );
+        } catch (IOException x) {
+            throw new GATKException("Couldn't read "+bam+": "+x.getMessage(), x);
+        }
+
+        // fill in reference bases
+        reads = reads.map(s -> AddContextDataToReadOptimizedFn.fillContext(rds, s));
+
+        logger.info("computing recal tables");
+
+        // run BaseRecalibratorEngine.
+        BaseRecalibratorEngineSpark recal = new BaseRecalibratorEngineSpark(readsHeaderView, refDictionaryView, bqsrArgs);
+        JavaRDD<RecalibrationTables> tables = reads.mapPartitions(s->recal.apply(s));
+
+        logger.info("merging recal tables");
+
+        RecalibrationTables table = tables.reduce( (a,b)->a.combine(b) );
```

Comment:
In `BaseRecalibratorSpark` we found that switching to a hierarchical reduce here sped up the reduce step by a huge amount. Recommend that you do the same in this version of the tool:

            final RecalibrationTables emptyRecalibrationTable = new RecalibrationTables(new StandardCovariateList(bqsrArgs, readsHeader));
            final RecalibrationTables table = tables.treeAggregate(emptyRecalibrationTable,
                RecalibrationTables::inPlaceCombine,
                RecalibrationTables::inPlaceCombine,
                Math.max(1, (int)(Math.log(tables.partitions().size()) / Math.log(2))));

See `BaseRecalibratorSparkFn` lines 48-52.